> Composing computer programs to solve scientific problems is like writing poetry. You must choose every word with care and link it with the other words in perfect syntax.
**James Lovelock**

**Why is programming so hard?**

**Difficulty 1**: Two Things at Once.

* You are not yet familiar with the syntax and semantics of the language you are working with—in this case, of the programming language Python.
* You do not know how to solve problems using a computer—similar to not knowing how to write poetry.

**Difficulty 2**: What is a Good Program?
* Problem solving

Writing a program as you figure out how to solve a problem means that you are *working at two levels at once*: **the problem-solving level and the programming level**.That is more difficult than doing things sequentially.You should sit down and think about the problem and how you want to solve it before you start writing the program.

> **Rule 1**: Think before you program!

 “What is the most important feature a program should have?” many answer, “It should run.” By “run,” they mean that the program executes and actually does something.

> **Rule 2**: A program is a human-readable essay on problem solving that also happens to execute on a computer.

**Computation** - is the manipulation of data by either humans or machines.

**Computer** is something that does computation.
* A computer should be able to accept input. What counts as input might vary among computers, but data must be able to enter the computer for processing.
* If a computer is defined by its ability to do computation, then any object that is a computer must be able to manipulate data.
* A computer must be able to output data. 

> A good workman is known by his tools.
**proverb**

Our first steps in programming are to learn the details, syntax and semantics of the Python programming language.

Before we get too far along, we want to emphasize something important. One of the best reasons to start learning programming using Python is that you can easily experiment with Python. That is, Python makes it easy to try something out and see the result. Anytime you have a question, simply try it out. Learning to experiment with a programming language is a very important skill, and one that seems hard for introductory students to pick up. So let’s add a new RULE.

> **Rule 3**: The best way to improve your programming and problem skills is to practice!

## Quickstart: The Circumference program

We want to calculate the circimference and area of a circle given its radius. The relevant mathematical formulas are:
* circumference = 2 * pi * radius
* ares = pi * radius ^ 2

In [ ]:
# Calculate the area and circumference of a circle from its radius.
# Step 1: Prompt for a radius.
# Step 2: Apply the area formula.
# Step 3: Print out the results.

import math
radius_str = input("Enter the radius of your circle:")

radius_int = int(radius_str)

circumference = 2 * math.pi * radius_int
area = math.pi * (radius_int ** 2)

print ("The cirumference is: ",circumference, \
       ", and the area is: ",area)


Enter the radius of your circle:25
The cirumference is:  157.07963267948966 , and the area is:  1963.4954084936207


**An Interactive Session**

An important feature of Python is that it is an *interpreted* language. By interpreted we mean that there is a program within Python called the interpreter that takes each line of Python code, one line at a time, and executes that code. This feature meas that we can try out lines of code one at a time by typing into the Python shell.

In [ ]:
import math
radius_str = input("Enter the radius of your circle:")

Enter the radius of your circle:25


In [ ]:
radius_str

'25'

In [ ]:
radius_int = int(radius_str)
radius_int

25

In [ ]:
math.pi

3.141592653589793

In [ ]:
circumference = 2 * math.pi * radius_int
area = math.pi * (radius_int ** 2)

In [ ]:
print ("The cirumference is: ",circumference, \
       ", and the area is: ",area)

The cirumference is:  157.07963267948966 , and the area is:  1963.4954084936207


## Parts of a Program

**Modules**
* A *module* contains a set of Python commands.
* A module can be stored as a file and *imported* into the Python shell. Ex: *import math*

**Statements** and **Expressions**

*Expression*: a combination of values and operations that creates a new value that we call a *return value* - i.e. the value returned by the operation(s).

*Statement*: doesn't return a value, but does perform some task. Some statements may control the flow of the program, and others might ask for resources.

In [ ]:
my_int = 5 # # statement, no return value but my int now has value 5 
my_int

5

In [ ]:
my_int + 5 # # expression, value associated with my int added to 5 

10

**Whitespace**

Python has the following rules about hot whitespace is used in a program:
* Whitespace is ignored within both expressions and statements. For example, Y=X+5 has exactly the same meaning as Y =    X +           5
* Leading whitespace, whitespace at the beginning of a line, defines indentation. Indentation plays a special role in Python.
* Blank lines are also considered to be whitespace, and the rule for blank lines is trivial: **blank lines are allowed anywhere and are ignored**.

**Indentation**

Indentation is used by all programmers to make code more readable. Programmers often indent code to indicate that the indented code *grouped together*, meaning those statements have some common purpose. However, indentation is treated uniquely in Python. Python *requires* it for grouping.

In [ ]:
my_int = 12
if my_int < 20:
    print('less than 20')
else:
    print('greather than 20')

less than 20


In [ ]:
for i in range(10):
    if i % 2 == 0:
        print(i)

0
2
4
6
8


**Continuation**

Long lines of code, those wider than the width of a reasonably sized editing window, can make reading code difficult. Because readability is very important (remember RULE 2), Python provides ways to make long lines more readable by splitting them. Such splitting is called a continuation.  You indicate a continuation by placing a backslash character **\** at the end of a line.

In [ ]:
print ("The cirumference is: ",circumference, ", and the area is: ",area)

The cirumference is:  157.07963267948966 , and the area is:  1963.4954084936207


In [ ]:
print ("The cirumference is: ",circumference, \
       ", and the area is: ",area)

The cirumference is:  157.07963267948966 , and the area is:  1963.4954084936207


**Comments**

A program is more than "just some code that does something". A program is a document that describes the thought process of its writer. Messy code implies messy thinking and is both difficult to work with and understand. Code also happens to be something that can run, but just because it can run does not make it a good program. Good programs can be read, just like any other essay. Comments are one important way to improve readability. Comments contribute nothing to the running of the program, because Python ignores them. In Python, anything following a pound character **#** is ignored on that line. However, comments are critical to the readability of the program. 

In [ ]:
# import package
import math

# get input from user
radius_str = input("Enter the radius of your circle:")

# convert string into integer
radius_int = int(radius_str)

Enter the radius of your circle:23


**Special Python Elements: Tokens**

As when learning any language, there are some details that are good to know before you can fully understand how they are used. Here we show you the special keywords, symbols, and characters that can be used in a Python program. These language elements are known generically as *tokens*. We don’t explain in detail what each one does, but it is important to note that Python has special uses for all of them.

*Keywords*

Keywords are special words in Python that cannot be used by you to name things. They indicate commands to the Python interpreter.
![](figures/python-keywords.png)

*Operators* 

Operators are special tokens (sequences of characters) that have meaning to the Python interpreter. Using them implies some operation, such as addition, subtraction, or something similar.
![](figures/python-operators.png)

*Punctuators and Delimiters*

Punctuators, a.k.a. delimiters, separate different elements in Python statements and expres- sions.
![](figures/python-punctuators.png)

**Naming Objects**

If writing a program is like writing an essay, then you might guess that the names you use in the program, such as the names of your variables, would help greatly in making  the program more readable. Therefore, it is important to choose names well.

1. Every name must begin with a letter or the underscore character _:
    * A numeral is not allowed as the first character.
    * Multiple-word names can be linked together using the underscore character _ — e.g. *this_is_a_variable_name_with_underscore*. A name starting with an underscore is often used by Python and Python programmers to denote a variable with special characteristics. You will see these as we go along, but for now it is best to not start variable names with an underscore until you understand what that implies.
2.	After the first letter, the name may contain any combination of letters, numbers, and underscores:
    * The name cannot be a keyword.
    * You cannot have any delimiters, punctuation, or operators in a name.
3.	A name can be of any length.
4.	UPPERCASE is different from lowercase:
    * my_name is different than my_Name or My_Name or My_name.


[Style Guide for Python Code](https://www.python.org/dev/peps/pep-0008)


> **Rule 4**:  A foolish consistency is the hobgoblin of little minds.

**Variables**

A *variable* is a name you create in your program to represent “something” in your program. That “something” can be one of many types of entities: a value, another program to run, a set of data, a file. For starters, we will talk about a variable as something that represents a value: an integer, a string, a floating-point number, and so on. We create variables with meaningful names, because the purpose of a good variable name is to make your code more readable. The variable name pi is a name most would recognize, and it is easier to both read and write than 3.1415926536. Once a variable is created, you can store, retrieve, or modify the data associated with that variable. Every time you use your variable in a program, its value is retrieved by Python and used in that variable’s place. Thus, a variable is really a way to make it easier to read the program you are writing.

How does Python associate the value of the variable with its name? In essence, Python makes a list of names (variables, but other names as well) that are being used right now in the Python interpreter. The interpreter maintains a special structure called a namespace to keep this list of names and their associated values. Each name in that list is associated with a value, and the Python interpreter updates both names and values during the course of its operation. The name associated with a value is an alias for the value, i.e., another name for it. Whenever a new variable is created, its name is placed in the list along with an association to a value. If a variable name already exists in the table, its association is updated.

In [ ]:
my_int = math.pi + 5

In this case, 5 is added to the value of variable associated with math.pi and the result of that expression is associated with the variable my int in the Python namespace. It is important to note that the value math.pi is not modified by this operation. In fact, assignment does not change any values on the right-hand side. Only a new association is created by assignment—the association with the name on the left-hand side.

The general form of the assignment statement is the same as in mathematics:

> left-hand side = right-hand side

Although we use the familiar equal sign (=) from mathematics, its meaning in pro- gramming languages is different! In math, the equal (=) indicates equality: what is on the left-hand side of the equal sign (=) has the same value as what is on the right-hand side. In Python, the equal sign (=) represents assignment. Assignment is the operation to associate a value with a variable. The left-hand side represents the variable name and the right-hand side represents the value. If the variable does not yet exist, then the variable is created and placed in the namespace otherwise, the variable’s value is updated to the value on the right-hand side. This notation can lead to some odd expressions that would not make much sense mathematically but make good sense from Python’s point of view, such as:

> my_var = my_var + 1


Evaluation of an assignment statement is a two-step process:
1.	Evaluate the expression on the right-hand side.
2.	Take the resulting value from the right-hand expression and associate it with the variable on the left-hand side (create the variable if it doesn’t exist; otherwise update it).

![](figures/namespace.png)

**Objects and Types**

In assignment, we associate a value with a variable. What exactly is that value? What information is important and useful to know about that value?

In Python, every “thing” in the system is considered to be an object. In Python, though, the word object has a very particular meaning. An object in Python has:
* An identity
* Some attributes
* Zero or more names

Whenever an object is created by Python, it receives an identification number. If you are ever interested in the number of any object, you can use the id function to discover its ID number. In general, the ID number isn’t very interesting to us, but that number is used by Python to distinguish one object from another. We will take a brief look at the ID here because it helps explain how Python manages objects.

Notice that in addition to the ID number, an object can also have a name, or even multiple names. This name is not part of the object’s ID but is used by us, the programmers, to make the code more readable. Python uses the namespace to associate a name (such as a variable name) with an object. Interestingly, multiple namespaces may associate different names with the same object!

Finally, every object has a set of attributes associated with it. Attributes are essentially information about the object. We will offer more insight into object attributes later, but the one that we are most interested in right now is an object’s type.

Knowing the type of an object informs Python (and us, the programmers) of two things:
* *Attributes* of the object tell us something about its ‘‘content.’’ For example, there are no decimal points in an integer object, and there are no letters in either an integer object or a float object.
* *Operations* we can perform on the object and the results they return. For example, we can divide two integer objects or two float objects, but the division operation makes no sense on a string object.


In [ ]:
a_int = 7
id(a_int)

140709943940096

In [ ]:
type(a_int)

int

In [ ]:
b_float = 2.5
id(b_float)

1877679067880

In [ ]:
type(b_float)

float

In [ ]:
a_int = b_float # associate a_int with value of b_float
type(a_int) # a_int contains a type float

float

In [ ]:
id(a_int) # a_int has the same ID as b_float

1877679067880

![](figures/namespace-before-after.png)

**Numbers**

*Integers*
The integer type is designated in Python as type int. The integer type corresponds to our mathematical notion of integer.

In [ ]:
012 # leading zero without letter is invalid

SyntaxError: invalid token (<ipython-input-28-530046ea2c0f>, line 1)

In [ ]:
125

125

*Floating-Point Numbers*
Floating-point or real numbers are designated in Python as type float. The floating-point type refers to noninteger numbers—numbers with decimal points. Floats are created either by typing the value, such as 25.678, or by using exponential notation, with the exponent represented by an “e.” The operators are, like integers.

Floats represent Real numbers, but only approximately. For example, what is the exact decimal representation of the operation 2.0/3.0? As you know, there is no exact decimal equivalent as the result is an infinite series: 2.0/3.0 = 0.666  Because a computer has a finite amount of memory, real numbers must be represented with approximations of their actual value. Look at the following exampes.

In [ ]:
2.0 / 3.0

0.6666666666666666

In [ ]:
1.1 + 2.2

3.3000000000000003

In [ ]:
0.1 + 0.1 + 0.1 - 0.3

5.551115123125783e-17

If you were to do the calculations yourself on paper, you would find that 1.1+2.2 is equal to 3.3. However, the session shows it is 3.3000000000000003. Same for the last addition. The result should be zero but Python returns a very small value instead. Approximations like this, if carried through multiple evaluations, can lead to significant differences than what is expected. Python does provide a module called the decimal module that provides more predictable, and controllable, results for floating-point numbers.

It is important to remember that floating-point values are approximate values, not exact, and that operations using floating-point values yield approximate values. Integers are exact, and operations on integers yield exact values.

Finally, unlike integers in Python, a leading 0 on a floating-point number carries no significance: 012. (notice the decimal point) is equivalent to 12.0.

In [ ]:
012.3

12.3

In [ ]:
0012.0

12.0

**Other Built-In Types**

* Boolean - True or False
* String - 'any text'
* List - [4, 3.26, 'Baku']
* Dictionary - {'Namiq' : '15-March-1989', 'Rasim' : '14-June-2008'}
* Tuple - (1, 2)
* Set - {'ag', 'qara', 'qirmizi'}

**Operators**

For every type there is a set of operations that can be performed on that type. Given the limited number of symbols available, some symbols are used for different purposes for different types. This is called operator **overloading**, a term that indicates that a symbol might have multiple meanings depending on the types of the values. For example, we saw earlier that the plus sign (+) performs different operations depending on whether the operands are integers or strings.

**Integer Operators**

Most of the operators for integers work exactly as you learned in elementary arithmetic class. The plus sign is used for addition, the minus is for subtraction, and the asterisk is for multiplication.

Division works as you would expect as well, but it is a little different from the other operators. In particular, the integers are not closed with respect to division. By closure, we mean that if you add two integers, the sum is an integer. The same is true for subtraction and multiplication. However, if you divide two integers, the result is not necessarily an integer. It could be an integer (say, 4/2) or it could be a float (4/3). Mathematically it is a rational number (a.k.a. fraction), but Python represents it as a float.

> For consistency’s sake, then, whenever you do division, regardless of the types of the operands, the type yielded is a **float**.

In [ ]:
a_int = 4
b_int = 2

a_int + b_int

6

In [ ]:
a_int - b_int

2

In [ ]:
a_int * b_int

8

In [ ]:
a_int / b_int

2.0

In [ ]:
result = a_int / b_int
type(result)

float

In [ ]:
a_int / 3

1.3333333333333333

In [ ]:
5 / 3

1.6666666666666667

In [ ]:
5 // 3 # integer division

1

In [ ]:
5 % 3 # integer remainder

2

In [ ]:
5.0 // 3.0 # integer division but as a float

1.0

In [ ]:
5.0 % 3.0 # integer remainder but as a float

2.0

In [ ]:
5.0 / 3.0

1.6666666666666667

In [ ]:
5 // 3.0 # when types are mixed, result is a float

1.0

![](figures/operators.png)

**Order of Operations and Parantheses**

The order of arithmetic operations in Python and most common programming languages is the same as the one you learned in arithmetic: multiplication and division before addition or subtraction. The term used to describe the ordering is *precedence*. In this case, we say that multiplication and division have greater precedence than addition or subtraction, so they are done first. Further, exponents have greater precedence than multiplication and division, also as in arithmetic. If operations have the same precedence—that is, multiplication and division—they are done left to right. Finally, as in arithmetic, parentheses can be used to override the precedence order and force some operations to be done before others, regardless of precedence. (from highest precedence (performed first) to lowest)
![](figures/precedence.png)

In [ ]:
2 + 3 - 4 # same precedence: left to right

1

In [ ]:
4 / 2 * 5 # same precedence: left to right

10.0

In [ ]:
2 + 3 * 5 # multiplication before addition

17

In [ ]:
(2 + 3) * 5 # parantheses force addition before multiplication

25

In [ ]:
2 + 3 * 5 ** 2 # exponents before multiplication before addition

77

In [ ]:
2 + 3 * 5 ** 2 - 1

76

In [ ]:
-4 + 2 # negation before addition and substruction

-2

**Augmented Assignment Operators: A Shortcut!**

Operations—especially groups of operations—that are used repeatedly are often provided with a shortcut, a simpler way of typing. Python, like other languages, has such shortcuts. They are not required, but they do make typing easier and, at some point, make the code shorter and easier to read.
![](figures/shortcuts.png)

> **Rule 5**: Test your code, often and thoroghly!

# Exercices

`Body mass index (BMI)` is a number calculated from a person’s weight and height. According to the Centers for Disease Control and Prevention, the BMI is a fairly reliable indicator of body fatness for most people. BMI does not measure body fat directly, but research has shown that BMI correlates to direct measures of body fat, such as underwater weighing and dual-energy X-ray absorptiometry.
The formula for BMI is:

> weight/ height ^ 2

where weight is in *kilograms* and height is in *meters*.
* Write a program that prompts for metric weight and height and outputs the BMI.
* Write a program that prompts for weight in pounds and height in inches, converts the values to metric, and then calculates the BMI.


`The Great Lakes are how big?`

The Great Lakes in the United States contain roughly 22% of the world’s fresh surface water (22,810 km3). It is hard to conceive how much water that is. Write a program to calculate how deep it would be if all the water in the Great Lakes were spread evenly across the 48 contiguous U.S. states. You will need to do some Internet research to determine the area of that region.


`Population estimation`
The U.S. Census provides information on its web page (http://www.census.gov) about the current U.S. population as well as approximate rates of change.

Three rates of change are provided:
* There is a birth every 7 seconds. There is a death every 13 seconds.
* There is a new immigrant every 35 seconds.
* These are obviously approximations of birth, death, and immigration rates, but they can assist in providing population estimates in the near term.

Write a program that takes years as input (as an integer) and prints out an estimated population (as an integer). Assume that the current population is 307,357,870, and assume that there are exactly 365 days in a year.

**Hint**: Note that the rate units are in seconds.
